src/anyp/Uri.cc

```diff
@@ -81,6 +81,66 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    const auto end = buf.end();
+    for (auto i = buf.begin(); i != end; ++i) {
+        const auto ch = *i;
+        if (ch != '%') {
+            output.append(ch);
+            continue;
+        }
+        // ch is '%', let's look ahead
+        ++i;
+        if (i == end) {
+            // '%' is the last character, so it's not encoded
+            output.append(ch);
+            break;
+        }
+        const auto ch1 = *i;
+        if (ch1 == '%') {
+            // met "%%"; output "%" and continue
+            output.append(ch);
+            continue;
+        }
+        const auto hex1 = HextDigitToInt(ch1);
+        if (hex1 == -1) {
+            // ch1 is not a hex digit, so it's not encoded
+            output.append(ch).append(ch1);
+            continue;
+        }
+        ++i;
+        if (i == end) {
+            // end is one-past-last, so ch2 is actually garbage
+            output.append(ch).append(ch1);
+            break;
+        }
+        const auto ch2 = *i;
+        const auto hex2 = HextDigitToInt(ch2);
+        if (hex2 == -1) {
+            // ch2 is not a hex digit, so it's not encoded
+            output.append(ch).append(ch1).append(ch2);
```

Comment:
The RFC reads:
`A percent-encoded octet is encoded as a character triplet`, or in BNF: `pct-encoded = "%" HEXDIG HEXDIG`, so the %4 should be copied over, and we are left with `%4A`.

The case of `%4%%41` is more interesting, because given that %% -> % is not mentioned in the RFC, we could say right now we are not compliant, and by the reading of the RFC we should output %4%A.

I think the key to decide what to do is in the last paragraph of [RFC 3986 section 2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4), which says:
```
Because the percent ("%") character serves as the indicator for
percent-encoded octets, it must be percent-encoded as "%25" for that
octet to be used as data within a URI
```
Which means that %4% is an invalid %-encoded string (as is %%).

Reply:
> The RFC reads: `A percent-encoded octet is encoded as a character triplet`, or in BNF: `pct-encoded = "%" HEXDIG HEXDIG`,

Yes.

> so the %4 should be copied over

Why?! I see no logical connection between the two statements and nothing in the URI grammar that allows us to "copy over" bare `%`. As I said [earlier](https://github.com/squid-cache/squid/pull/1626#discussion_r1436201046) (and you may have implied in your "the key to decide" response quoted below), the grammar prohibits `%` outside of `pct-encoded` triplets. If those bare `%` are banned by the grammar, why should we copy them instead of rejecting malformed input?


> The case of `%4%%41` is more interesting ... by the reading of the RFC we should output %4%A.

Why?! I see no sequence of URI grammar rules that would yield that output for that input. What sequence produces that output?


> I think the key to decide what to do is in the last paragraph of [RFC 3986 section 2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4), which says:
> 
> ```
> Because the percent ("%") character serves as the indicator for
> percent-encoded octets, it must be percent-encoded as "%25" for that
> octet to be used as data within a URI
> ```

... and that prose is implied by the RFC grammar rules AFAICT -- bare `%` cannot appear anywhere except as the indicator for percent-encoded octets: `unreserved`, `sub-delims`, and other URI grammar rules do not allow `%` characters.
 
> Which means that %4% is an invalid %-encoded string (as is %%).

Agreed. And `%X` and `%AX`.